pkg/module/sync/runner.go

```diff
@@ -2,8 +2,270 @@ package sync
+			// Due to module template visibility change, some module previously deployed should be removed.
+			if errors.Is(module.Template.Err, channel.ErrTemplateNotAllowed) {
+				results <- r.deleteManifest(ctx, module)
+				return
+			}
+			// ModuleInStatus template in other error status should be ignored.
+			if module.Template.Err != nil {
+				results <- nil
+				return
+			}
+			if err := r.updateManifests(ctx, kyma, module); err != nil {
+				results <- fmt.Errorf("could not update module %s: %w", module.GetName(), err)
+				return
+			}
+			module.Logger(baseLogger).V(log.DebugLevel).Info("successfully patched module")
+			results <- nil
+		}(module)
+	}
+	var errs []error
+	for i := 0; i < len(modules); i++ {
+		if err := <-results; err != nil {
+			errs = append(errs, err)
+		}
+	}
+	ssaFinish := time.Since(ssaStart)
+	if len(errs) != 0 {
+		errs = append(errs, fmt.Errorf("ServerSideApply failed (after %s)", ssaFinish)) //nolint:goerr113
+		return errors.Join(errs...)
+	}
+	baseLogger.V(log.DebugLevel).Info("ServerSideApply finished", "time", ssaFinish)
+	return nil
+}
+
+func (r *Runner) getModule(ctx context.Context, module client.Object) error {
+	err := r.Get(ctx, client.ObjectKey{Namespace: module.GetNamespace(), Name: module.GetName()}, module)
+	if err != nil {
+		return fmt.Errorf("failed to get module by name-namespace: %w", err)
+	}
+	return nil
+}
+
+func (r *Runner) updateManifests(ctx context.Context, kyma *v1beta2.Kyma,
+	module *common.Module,
+) error {
+	if err := r.setupModule(module, kyma); err != nil {
+		return err
+	}
+	obj, err := r.converter.ConvertToVersion(module.Manifest, r.versioner)
+	if err != nil {
+		return fmt.Errorf("failed to convert object to version: %w", err)
+	}
+	manifestObj, ok := obj.(*v1beta2.Manifest)
+	if !ok {
+		return commonerrs.ErrTypeAssert
+	}
+	if err := r.Patch(ctx, manifestObj,
+		client.Apply,
+		client.FieldOwner(kyma.Labels[v1beta2.ManagedBy]),
+		client.ForceOwnership,
+	); err != nil {
+		return fmt.Errorf("error applying manifest %s: %w", client.ObjectKeyFromObject(module), err)
+	}
+	module.Manifest = manifestObj
+
+	return nil
+}
+
+func (r *Runner) deleteManifest(ctx context.Context, module *common.Module) error {
+	err := r.Delete(ctx, module.Manifest)
+	if util.IsNotFound(err) {
+		return nil
+	}
+	return fmt.Errorf("failed to delete manifest: %w", err)
+}
+
+func (r *Runner) setupModule(module *common.Module, kyma *v1beta2.Kyma) error {
+	// set labels
+	module.ApplyLabelsAndAnnotations(kyma)
+	refs := module.GetOwnerReferences()
+	if len(refs) == 0 {
+		// set owner reference
+		if err := controllerutil.SetControllerReference(kyma, module.Manifest, r.Scheme()); err != nil {
+			return fmt.Errorf("error setting owner reference on component CR of type: %s for resource %s %w",
+				module.GetName(), kyma.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (r *Runner) SyncModuleStatus(ctx context.Context, kyma *v1beta2.Kyma, modules common.Modules,
+	metrics ModuleMetrics,
```

Comment:
Now since you did this great refactoring we could add `ModuleMetrics` as a field to `Runner` and pass it via `func New(*)` as well. Since it's used by 2 public methods here.  WDYT?

Reply:
I think we should not introduce `ModuleMetrics` to `Runner`, as I mentioned in our discussion, the metrics to me are more low-level utility and usually address cross-cutting concerns, it should be flexible enough to pass into the actual place as composition, but should not belong to any specific business object. It's similar to using a struct to carry logger, to me, it's all anti-pattern. Why does this runner need a metric?

Although it's convent to carry those utility objects, it increases code rigidity.

Reply:
Ok let's try that approach then, fine by me 👍